artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +82,209 @@ def injection_status_cb(channel, override, value):
+        assert final_values[urukul.chip_select - 4] == ftw == target_ftw
+
+    @async_test
+    async def test_ftw_int32(self):
+        target_ftws = {
+            0: 0xaabbccdd,
+            1: 0xabababab,
+            2: 0xabcdabcd,
+            3: 0x900f009
+        }
+        ftws = {}
+        notifications_out = []
+        async with self.open_comm_session(notifications_out):
+            for name, urukul in self.urukuls_all().items():
+                idx = urukul.chip_select - 4
+                self.kernel.write_raw(urukul, target_ftws[idx])
+                ftws[idx] = self.kernel.read_raw(urukul)[0]
+        final_values = {probe: value for _, probe, value in notifications_out}
+        assert final_values == ftws == target_ftws
+
+    @async_test
+    async def test_ftw_int48(self):
+        self.ensure_ad9912_only()
+
+        target_ftws = {
+            0: 0xffeeffeeffee,
+            1: 0xfeedfeedfeed,
+            2: 0xabababababab,
+            3: 0xf00ba12ba2
+        }
+        ftws = {}
+        notifications_out = []
+        async with self.open_comm_session(notifications_out):
+            for name, urukul in self.urukuls["AD9912"].items():
+                idx = urukul.chip_select - 4
+                self.kernel.write_raw(urukul, target_ftws[idx])
+                ftws[idx] = self.kernel.read_raw(urukul)[0]
+        final_values = {probe: value for _, probe, value in notifications_out}
+        assert final_values == ftws == target_ftws
+
+    @async_test
+    async def test_frequency(self):
+        target_freqs = {
+            0: 15 * MHz,
+            1: 42.5 * MHz,
+            2: 98.765 * MHz,
+            3: 128.128 * MHz
+        }
+        freqs = {}
+        notifications_out = []
+        async with self.open_comm_session(notifications_out):
+            for name, urukul in self.urukuls_all().items():
+                idx = urukul.chip_select - 4
+                self.kernel.write(urukul, target_freqs[idx])
+                freqs[idx] = self.kernel.read(urukul)[0]
+        for key, value in freqs.items():
+            assert numpy.isclose(value, target_freqs[key], rtol=1e-06)
+
+    @async_test
+    async def test_frequency_with_turns(self):
+        target_freqs = {
+            0: 15 * MHz,
+            1: 42.5 * MHz,
+            2: 98.765 * MHz,
+            3: 128.128 * MHz
+        }
+        freqs = {}
+        notifications_out = []
+        async with self.open_comm_session(notifications_out):
+            for name, urukul in self.urukuls_all().items():
+                idx = urukul.chip_select - 4
+                self.kernel.write(urukul, target_freqs[idx], 123.4)
+                freqs[idx] = self.kernel.read(urukul)[0]
+        for key, value in freqs.items():
+            assert numpy.isclose(value, target_freqs[key], rtol=1e-06)
+
+    @async_test
+    async def test_ad9910_set_all(self):
+        self.ensure_ad9910_only()
```

Comment:
only?

Reply:
changed its name to reflect what is it supposed to do